static/app/views/starfish/utils/zeroFillSeries.tsx

```diff
@@ -85,7 +87,12 @@ export function zeroFillSeries(
 }
 
 function roundUpToNearest12HourInterval(time: moment.Moment) {
-  return roundDownToNearest12HourInterval(time.clone().add(12, 'hour').subtract(1, 'ms'));
+  return roundDownToNearest12HourInterval(
+    time
+      .clone()
+      .add(12, 'hour')
+      .subtract(TWELVE_HOURS_IN_MS - 1, 'ms')
```

Comment:
ugh my bad, I should have just done `.add(1, 'ms')` as you pointed out. previously, we were doing an equivalent to `.add(~12 hours)` and now we're doing `.add(1 ms)` and then rounding down to the nearest 12 hour so that we don't potentially skip a bucket (this only matters if data is sparse). here's an example:


let's say your start time was `May 23 15:43:06.537`
in the old code, we'd be adding _just_ under 12 hours hours to bring it to `May 24 03:43:06.536` and then
rounding it down to the nearest 12 hour `May 24 00:00:00.000`

but since start time is prior to that, the series returned from clickhouse would have buckets starting at `May 23 12:00:00.000` (because of `toStartOfInterval`)
which is normally fine, but if data is sparse, and some series don't have that starting bucket, with the zerofilling, their "first datum" would be `May 24 00:00:00.000`
We want the sparse series to start at `May 23 12:00:00.000`

Which is why, instead of adding ~12 hours, we'd want to add 1 ms and round down to the nearest 12 hours
@gggritso

Reply:
actually, I think we can just do a straight up rounddown and never round up?

Reply:
🤔 yeah it sounds like rounding down is the right thing here